pkg/server/bootstrap/api-export-identity_controller.go

```diff
@@ -0,0 +1,255 @@
+	groupResourceExportNames map[schema.GroupResource]string,
+	kubeClusterClient kubernetes.ClusterInterface,
+	remoteShardApiExportInformer apisinformers.APIExportInformer,
+	configMapInformer coreinformers.ConfigMapInformer,
+) (*controller, error) {
+	queue := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), ControllerName)
+
+	c := &controller{
+		queue: queue,
+
+		kubeClusterClient: kubeClusterClient,
+
+		configMapIndexer:             configMapInformer.Informer().GetIndexer(),
+		remoteShardApiExportsIndexer: remoteShardApiExportInformer.Informer().GetIndexer(),
+
+		groupExportNames:         groupExportNames,
+		groupResourceExportNames: groupResourceExportNames,
+	}
+
+	interestingAPIExportNames := sets.NewString()
+	for _, name := range groupExportNames {
+		interestingAPIExportNames.Insert(name)
+	}
+	for _, name := range groupResourceExportNames {
+		interestingAPIExportNames.Insert(name)
+	}
+	c.interestingAPIExportNames = sets.NewString(interestingAPIExportNames.List()...)
+	remoteShardApiExportInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: func(obj interface{}) bool {
+			key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+			if err != nil {
+				return false
+			}
+			clusterName, _ := clusters.SplitClusterAwareKey(key)
+			if clusterName != tenancyv1alpha1.RootCluster {
+				return false
+			}
+			switch t := obj.(type) {
+			case *apisv1alpha1.APIExport:
+				return interestingAPIExportNames.Has(t.Name)
+			}
+			return false
+		},
+		Handler: cache.ResourceEventHandlerFuncs{
+			AddFunc:    func(obj interface{}) { c.queue.Add(workKey) },
+			UpdateFunc: func(old, new interface{}) { c.queue.Add(workKey) },
+			DeleteFunc: func(obj interface{}) { c.queue.Add(workKey) },
+		},
+	})
+
+	configMapInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: func(obj interface{}) bool {
+			key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+			if err != nil {
+				return false
+			}
+			_, clusterAwareName, err := cache.SplitMetaNamespaceKey(key)
+			if err != nil {
+				runtime.HandleError(err)
+				return false
+			}
+			clusterName, _ := clusters.SplitClusterAwareKey(clusterAwareName)
+			if clusterName != configshard.SystemShardCluster {
+				return false
+			}
+			switch t := obj.(type) {
+			case *corev1.ConfigMap:
+				return t.Name == configMapName
+			}
+			return false
+		},
+		Handler: cache.ResourceEventHandlerFuncs{
+			AddFunc:    func(obj interface{}) { c.queue.Add(workKey) },
+			UpdateFunc: func(old, new interface{}) { c.queue.Add(workKey) },
+			DeleteFunc: func(obj interface{}) { c.queue.Add(workKey) },
+		},
+	})
+	return c, nil
+}
+
+// Start starts the controller, which stops when ctx.Done() is closed.
+func (c *controller) Start(ctx context.Context, numThreads int) {
+	defer runtime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	klog.Infof("Starting %s controller", ControllerName)
```

Comment:
Do you mind using logging from e.g. anything in apis/tenancy to follow the structured/contextual pattern?

Reply:
use `klog.InfoS`, right ?

Reply:
Not quite - see

https://github.com/kcp-dev/kcp/blob/3af88fc9027fea16972ddfd4a1a76a0010255f38/pkg/reconciler/apis/apibinding/apibinding_controller.go#L134

https://github.com/kcp-dev/kcp/blob/3af88fc9027fea16972ddfd4a1a76a0010255f38/pkg/reconciler/apis/apibinding/apibinding_controller.go#L339-L342

https://github.com/kcp-dev/kcp/blob/3af88fc9027fea16972ddfd4a1a76a0010255f38/pkg/reconciler/apis/apibinding/apibinding_controller.go#L364-L366

https://github.com/kcp-dev/kcp/blob/3af88fc9027fea16972ddfd4a1a76a0010255f38/pkg/reconciler/apis/apibinding/apibinding_controller.go#L382

https://github.com/kcp-dev/kcp/blob/3af88fc9027fea16972ddfd4a1a76a0010255f38/pkg/reconciler/apis/apibinding/apibinding_controller.go#L400-L401

@stevekuznetsov do we have this doc'd?

Reply:
Not yet, but every other reconciler uses it, so it should be easy enough to copy this time ;)